sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+  // matches list of used aspects. There are no virtual member functions defined
+  // in the default set, which means that no extra optional kernel features
+  // requirements were attached to the kernel.
+});
+----
+
+Submitting a kernel with `calls_indirectly` property, which includes virtual
+member functions that use optional kernel features to a device that doesn't
+support them, should result in an exception at runtime, similar to how it is
+defined by the core SYCL specification.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_A;
+struct set_B;
+
+struct Foo {
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void foo() {
+    double d = 3.14;
+  }
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void bar() {}
+};
+
+sycl::queue q(/* device selector returns a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+q.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {
+  // Exception is expected to be thrown, because target device doesn't support
+  // fp64 aspect and it is required by 'Foo::foo' which is included into 'set_A'
+});
+
+q.single_task(syclext::properties{syclext::calls_indirectly<set_B>}, [=] {
+  // No exceptions are expected, because 'set_B' doesn't bring any requirements
+  // for optional kernel features.
+});
+----
+
+An implementation may not raise a compile time diagnostic or a run time
+exception merely due to speculative compilation of a virtual member function for
+a device when the application does not specify a use of virtual member functions
+through the correponding properties.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct Foo {
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {
+  double d = 3.14;
+}
+
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void bar() {}
+};
+
+sycl::queue q(/* device selector choosing a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+auto *Storage = sycl::malloc_device<Foo>(1, q);
+
+q.single_task([=] {
+  // The kernel is not submitted with 'calls_indirectly' property and therefore
+  // it is not considered to be using any of virtual member functions of 'Foo'.
+  // This means that the object of 'Foo' can be successfully created by this
+  // kernel, regardless of whether a target device supports 'fp64' aspect which
+  // is used by 'Foo::foo'.
+  // No exceptions are expected to be thrown.
+  new (Storage) Foo;
+});
+----
+
+==== Interaction with `reqd_sub_group_size` attribute
+
+The `reqd_sub_group_size` attribute is a bit of a special case comparing to
+other optional kernel features, because it requires to compile a kernel in a
+certain way, which may require special handling for all functions which are
+called from it.
+
+When the same helper function is called from two or more kernels
```

Comment:
I think this is the first occurrence of "helper function", and I'm not sure what it means.

Reply:
It was intended as a reference to the function from the previous paragraph, which is called from two kernels with different `reqd_sub_group_size` attributes. I've rewrote the paragraph to avoid using that undefined reference in 11f3a7d520fc29174fd21055704ecc97bd051c12